File: montecito/src/main/java/app/storkecentral/montecito/service/RinconService.java
package app.storkecentral.montecito.service;

import app.storkecentral.montecito.Config;
import app.storkecentral.montecito.model.Service;
import com.google.api.client.http.*;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.gson.Gson;

import java.io.IOException;
import java.net.ConnectException;
import java.util.HashMap;
import java.util.Map;

public class RinconService {

    public static Service service = new Service();

    static Gson gson = new Gson();

    static int retries = 0;

    public static void register() {
        try {
            Service montecito = new Service();
            montecito.setName("Montecito");
            montecito.setUrl("http://montecito:" + Config.PORT);
            montecito.setPort(Config.PORT);
            montecito.setStatusEmail(Config.STATUS_EMAIL);
            montecito.setVersion(Config.VERSION);

            HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
            HttpRequest request = requestFactory.buildPostRequest(
//                    new GenericUrl("http://localhost:" + Config.RINCON_PORT + "/services"),
                    new GenericUrl("http://rincon:" + Config.RINCON_PORT + "/services"),
                    ByteArrayContent.fromString("application/json", gson.toJson(montecito))
            );
            HttpResponse response = request.execute();
            if (response.getStatusCode() == 200) {
                System.out.println("Registered Montecito w/ Rincon!");
                registerRoutes();
            }
        } catch (IOException e) {
            System.out.println(e.getMessage());
            if (retries < 15) {
                retries++;
                System.out.println("Retrying Rincon connection attempt in 5s...");
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException ex) {
                    ex.printStackTrace();
                }
                register();
            }
            else {
                System.out.println("Failed to connect after 15 attempts, terminating program...");
                System.exit(100);
            }
        }
    }

    public static void registerRoutes() {
        try {
            HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory();
            HttpRequest request = requestFactory.buildPostRequest(
//                    new GenericUrl("http://localhost:" + Config.RINCON_PORT + "/routes"),
                    new GenericUrl("http://rincon:" + Config.RINCON_PORT + "/routes"),
                    ByteArrayContent.fromString("application/json", "{\n" +
                            "  \"route\": \"/montecito\",\n" +
                            "  \"service_name\": \"montecito\"\n" +
                            "}")
            );
            HttpResponse response = request.execute();
            if (response.getStatusCode() == 200) {
                System.out.println("Registered routes w/ Rincon!");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: arguello/service/rincon_service.go
package service

import (
	"arguello/config"
	"bytes"
	"encoding/json"
	"net/http"
	"os"
	"strconv"
	"time"
)

var rinconRetries = 0
var rinconHost = "http://rincon"

func RegisterRincon() {
	var portInt, _ = strconv.Atoi(config.Port)
	rinconBody, _ := json.Marshal(map[string]interface{}{
		"name":         "Lacumbre",
		"version":      config.Version,
		"url":          "http://lacumbre:" + config.Port,
		"port":         portInt,
		"status_email": config.StatusEmail,
	})
	responseBody := bytes.NewBuffer(rinconBody)
	_, err := http.Post(rinconHost+":"+config.RinconPort+"/services", "application/json", responseBody)
	if err != nil {
		if rinconRetries < 15 {
			rinconRetries++
			if rinconRetries%2 == 0 {
				rinconHost = "http://localhost"
				println("failed to register with rincon, retrying with \"http://localhost\" in 5s...")
			} else {
				rinconHost = "http://rincon"
				println("failed to register with rincon, retrying with \"http://rincon\" in 5s...")
			}
			time.Sleep(time.Second * 5)
			RegisterRincon()
		} else {
			println("failed to register with rincon after 15 attempts, terminating program...")
			os.Exit(100)
		}
	} else {
		println("Registered service with Rincon!")
		RegisterRinconRoute("/arguello")
		RegisterRinconRoute("/maps/buildings")
	}
}

func RegisterRinconRoute(route string) {
	rinconBody, _ := json.Marshal(map[string]string{
		"route":        route,
		"service_name": "Lacumbre",
	})
	responseBody := bytes.NewBuffer(rinconBody)
	_, err := http.Post(rinconHost+":"+config.RinconPort+"/routes", "application/json", responseBody)
	if err != nil {
	}
	println("Registered route " + route)
}
